crates/evm/src/execute.rs

```diff
@@ -2,12 +2,73 @@
 
 use reth_primitives::{BlockNumber, BlockWithSenders, Receipt, Receipts, Request, Requests, U256};
 use reth_prune_types::PruneModes;
-use revm::db::BundleState;
-use revm_primitives::db::Database;
+use revm::{db::BundleState, Context, DatabaseCommit, Evm};
+use revm_primitives::{db::Database, EVMError, ResultAndState};
 
 pub use reth_execution_errors::{BlockExecutionError, BlockValidationError};
 pub use reth_storage_errors::provider::ProviderError;
 
+/// An EVM executor trait that executes an input (e.g. transaction and height) and produces an output
+/// (e.g. state changes).
+///
+/// The executor cannot commit the state changes to the database directly, see [`EvmCommitter`].
+pub trait EvmExecutor {
+    /// The environment for the executor.
+    type Env;
+    /// The output produced by the executor.
+    type Output;
+    /// The error produced by the executor.
+    type Error;
+
+    /// Runs the executor with the current env.
+    fn execute(&mut self) -> Result<Self::Output, Self::Error>;
+
+    /// Returns the current env of the executor.
+    fn env(&self) -> &Self::Env;
```

Comment:
we also need env_mut, because we need to support this:

https://github.com/paradigmxyz/reth/blob/b7ce1de00eaf2c793dc967817e5b969948e23800/crates/rpc/rpc/src/eth/api/transactions.rs#L614-L625